### PR TITLE
Feature coalescing support

### PR DESCRIPTION
**Pull request by kylehuntsman**
Added coalescing support to save resources. Coalescing time window can be configured in the config in the "coalescingPeriod" property. Duplicate requests that are fired while a job is building will be thrown away and redirected to the running job. Duplicate requests that are fired during the coalescing period will also be thrown away and redirected to the original job.

### CODE REVIEW

**Comment by kylehuntsman** [comment]
Hold off on review until tests have been made.

**Comment by kylehuntsman** [comment]
Added integration test.

File: lib/datastore.js [hunk 1]
```diff
@@ -14,14 +14,13 @@ module.exports = Datastore;
  */
 function Datastore() {
   // TODO: Switch this to something that can manage the size, evict old, etc. LRU maybe?
-  this.cache = {};
+  this.cache = [];
```
Comment (sukrit007) [inline lib/datastore.js]: Clarification:  Any reason this needs to be an array instead of object ?

**Comment by sukrit007** [comment]
@FunnyGopher  Looks like there are some conflicts (possibly due to other branch that got merged with develop). Could you merge develop  onto this branch and resolve conflicts ?

File: lib/datastore.js (hunk 1 above)
Comment (kylehuntsman) [inline lib/datastore.js]: Oops. That was never reverted back. It should be an object.

**Comment by kylehuntsman** [comment]
Thanks for pointing that out. I need to go back and revert some of the code.

File: lib/factory.js [hunk 2]
```diff
@@ -425,3 +441,89 @@ Factory.prototype.createJob =  function createJob(req, res, context, next) {
   }
 };
 
+/**
+ * Returns an array of job objects
+ * @returns [] of jobs
+ */
+Factory.prototype.getJobs = function getJobs() {
+  var datastore = this.jobs.get();
```
Comment (sukrit007) [inline lib/factory.js]: **Suggestion**: Possibly use https://lodash.com/docs#values

File: lib/factory.js [hunk 3]
```diff
@@ -425,3 +441,89 @@ Factory.prototype.createJob =  function createJob(req, res, context, next) {
...
+/**
+ * Checks for duplicate requests and coalesces them if found
+ * @param res Response object
+ * @param context Job context holding job parameters (with defaults applied)
+ * @returns The original job {} if duplicates found, false if not
+ */
+Factory.prototype.coalesceRequests = function coalesceRequests(res, context) {
+  // Check for duplicate context in an array of jobs
+  function checkIfDuplicate(jobs) {
+    for (var j = 0; j < jobs.length; j++) {
```
Comment (sukrit007) [inline lib/factory.js]: **Suggestion**: Possibly use _.find  . See https://lodash.com/docs#find

**Pull request merged**